src/utils/meetup-events.ts

```diff
@@ -27,3 +48,77 @@ export function computeMeetupEventsApiResponse(
 
 export const pastEventsPageSise =
   parseInt(import.meta.env.PUBLIC_MEETUP_PAGE_SIZE, 10) || 10;
+
+/**
+ * Computes the 3rd Wednesday of the month of the input date
+ * and returns it as a Date.
+ * @param date input date.
+ */
+export function computeThirdWednesdayOfMonth(date: Date | number): Date {
+  const firstDayOfNextMonth = startOfMonth(date);
+
+  if (isWednesday(firstDayOfNextMonth)) {
+    return nextWednesday(nextWednesday(firstDayOfNextMonth));
+  }
+
+  return nextWednesday(nextWednesday(nextWednesday(firstDayOfNextMonth)));
+}
+
+export function computeScheduledEvents(
+  scheduledUpcomingEvents: MeetupEventType[] | undefined | null
+): ScheduledEvent[] {
+  return (
+    scheduledUpcomingEvents?.map(
+      (data): ScheduledEvent => ({
+        type: 'scheduled',
+        data,
+        epochTimeMs: new Date(data.dateTime).getTime(),
+      })
+    ) || []
+  );
+}
+
+/**
+ * Given a list of scheduled events, returns a list of scheduled and placeholder
+ * events sorted by date in ascending order.
+ * @param scheduledUpcomingEvents Scheduled events
+ * @param curentEpochTimeMs built time epoch time i.e. Date.now().
+ * @see {@link UpcomingEvent}
+ */
+export function computeUpcomingEvents(
+  scheduledUpcomingEvents: MeetupEventType[] | undefined | null,
+  curentEpochTimeMs: number
+): UpcomingEvent[] {
+  const scheduledEvents = computeScheduledEvents(scheduledUpcomingEvents);
+
+  const meetupOfThisMonth = computeThirdWednesdayOfMonth(curentEpochTimeMs);
+  const isNowBeforeMeetup =
+    !isSameDay(curentEpochTimeMs, meetupOfThisMonth) &&
+    isBefore(curentEpochTimeMs, meetupOfThisMonth);
+
+  const placeholderMonths = isNowBeforeMeetup ? [0, 1, 2, 3] : [1, 2, 3];
+
+  const placeholderEvents = placeholderMonths
+    .map((month): PlaceholderEvent => {
+      const date = computeThirdWednesdayOfMonth(
+        addMonths(new Date(curentEpochTimeMs), month)
+      );
+
+      return {
+        type: 'placeholder',
+        date,
+        epochTimeMs: date.getTime(),
+      };
+    })
+    .filter(
+      (placeholder) =>
+        !scheduledEvents.some(
+          (scheduled) =>
+            isSameDay(scheduled.epochTimeMs, placeholder.epochTimeMs) // filter out placeholder if overlaps with a scheduled event
+        )
+    );
+
+  return (scheduledEvents as UpcomingEvent[])
+    .concat(placeholderEvents)
+    .sort((a, b) => a.epochTimeMs - b.epochTimeMs);
+}
```

Comment:
Looks like the logic is starting to be complex enough to justify some tests.

Do you have opinions about it?

If you agree I can start by creating an issue

Reply:
- Added vitest as test runner.
- Tests now run on CI
- added [meetup-events.test.ts](https://github.com/Roma-JS/roma-js-on-astro/blob/24a9b8207c5ca7d555be5bac8c354f9d47c128e8/src/utils/tests/meetup-events.test.ts) tests.

Reply:
Awesome!